share-coordinator/src/test/java/org/apache/kafka/coordinator/share/ShareCoordinatorServiceTest.java

```diff
@@ -1980,6 +2020,121 @@ public void testColdPartitionSnapshotTaskPeriodicityWithSomeFailures() throws Ex
         service.shutdown();
     }
 
+    @Test
+    public void testPeriodicJobDoNotRunWhenShareGroupsDisabled() throws InterruptedException {
+        CoordinatorRuntime<ShareCoordinatorShard, CoordinatorRecord> runtime = mockRuntime();
+        PartitionWriter writer = mock(PartitionWriter.class);
+        MockTime time = new MockTime();
+        MockTimer timer = spy(new MockTimer(time));
+
+        Metrics metrics = new Metrics();
+
+        ShareCoordinatorService service = spy(new ShareCoordinatorService(
+            new LogContext(),
+            ShareCoordinatorTestConfig.testConfig(),
+            runtime,
+            new ShareCoordinatorMetrics(metrics),
+            time,
+            timer,
+            writer,
+            () -> false // So that the feature config is used.
+        ));
+
+        MetadataImage mockedImage = mock(MetadataImage.class, RETURNS_DEEP_STUBS);
+        MetadataDelta delta = mock(MetadataDelta.class);
+        ShareVersion shareVersion = mock(ShareVersion.class);
+        when(shareVersion.supportsShareGroups()).thenReturn(false);
+        when(mockedImage.features().finalizedVersions().getOrDefault(eq(ShareVersion.FEATURE_NAME), anyShort())).thenReturn((short) 0);
+
+        // Prune job.
+        when(runtime.scheduleWriteOperation(
+            eq("write-state-record-prune"),
+            any(),
+            any(),
+            any()
+        )).thenAnswer(inv -> {
+            service.onNewMetadataImage(mockedImage, delta);
+            return CompletableFuture.completedFuture(Optional.empty());
+        });
+
+        // Snapshot job.
+        when(runtime.scheduleWriteAllOperation(
+            eq("snapshot-cold-partitions"),
+            any(),
+            any()
+        )).thenAnswer(inv -> {
+            service.onNewMetadataImage(mockedImage, delta);
+            return List.of();
+        });
+
+        assertTrue(service.shouldRunPruneJob());
+
+        service.startup(() -> 1);
+        verify(timer, times(2)).add(any()); // Timer task added twice (prune, snapshot).
+        timer.advanceClock(30001L); // Will cause task execution.
+
+        verify(runtime, times(1)).scheduleWriteOperation(
+            eq("write-state-record-prune"),
+            any(),
+            any(),
+            any()
+        );
+
+        verify(runtime, times(1)).scheduleWriteAllOperation(
+            eq("snapshot-cold-partitions"),
+            any(),
+            any()
+        );
+
+        // New metadata image will have been served.
```

Comment:
If I'm reading this correctly, there will be an initial run of the periodic jobs, even if the feature and config are both disabled. That seems a bit surprising.

Reply:
The `oneNewMetadataImage` cannot be called unless the `service.startUp` is done (it will throw Error). Hence, at this point we cannot determine whether feature flag is up or not.
One way would be to move the invocation to `onNewMetadataImage` itself.